File: iOS/PengChauFerry/ui/FerryViewModel.swift
//
//  FerryViewModel.swift
//  PengChauFerry
//
//  Created by Lionello Lunesu on 7/3/2020.
//  Copyright © 2020 Lionello Lunesu. All rights reserved.
//

import Foundation
import Combine

final class FerryViewModel: ObservableObject {
    private let ferryRepository: FerryRepository
    private let holidayRepository: HolidayRepository

    init(ferryRepository: FerryRepository, holidayRepository: HolidayRepository) {
        self.ferryRepository = ferryRepository
        self.holidayRepository = holidayRepository
    }

    convenience init() {
        self.init(ferryRepository: FerryRepository(), holidayRepository: HolidayRepository())
    }

    struct State {
        let ferries: [Ferry]
        let from: FerryPier
        let day: FerryDay
    }

    @Published private(set) var state: State?
    private(set) var time = Timer.publish(every: 1, on: .main, in: .common).autoconnect()

    private var _time: LocalDateTime {
        LocalDateTime.now()
    }

    private func updateState(from: FerryPier, dow: FerryDay, autoRefresh: Bool, filtered: Bool) {
        var ferries = ferryRepository.getFerries(from: from, dow: dow)
        if autoRefresh && (ferries.isEmpty || holidayRepository.shouldRefresh()) {
            refreshAndUpdate(from: from)
        } else {
            if filtered {
                let now = _time.toLocalTime()
                ferries = ferries.filter { $0.time >= now }
            }
            self.state = State(ferries: ferries, from: from, day: dow)
        }
    }

    private var today: LocalDate {
        _time.toLocalDate()
    }

    func getDay(date: LocalDate) -> FerryDay {
        return holidayRepository.getHoliday(day: date) ? FerryDay.Holiday : FerryDay.fromDate(date)
    }

    func switchPier(_ pier: FerryPier) {
        updateState(from: pier, dow: getDay(date: today), autoRefresh: true, filtered: true)
    }

    private func refreshAndUpdate(from: FerryPier?) {
        let group = DispatchGroup()

        group.enter()
        ferryRepository.refresh {
            group.leave()
        }

        group.enter()
        holidayRepository.refresh {
            group.leave()
        }

        group.notify(queue: .main) {
            if let it = from {
                self.updateState(from: it, dow: self.getDay(date: self.today), autoRefresh: false, filtered: true)
            }
        }
    }

    func refresh() {
        refreshAndUpdate(from: state?.from)
    }

    func fetchAll() {
        if let it = state {
            updateState(from: it.from, dow: it.day, autoRefresh: true, filtered: false)
        }
    }

    func toggleHoliday() -> Bool {
        let today = self.today
        let isHoliday = getDay(date: today) == .Holiday
        holidayRepository.setHoliday(day: today, isHoliday: !isHoliday)
        if let it = state {
            updateState(from: it.from, dow: getDay(date: today), autoRefresh: true, filtered: true)
        }
        return !isHoliday
    }

}

File: Android/app/src/main/java/com/lunesu/pengchauferry/ui/ferry/FerryViewModel.kt
package com.lunesu.pengchauferry.ui.ferry

import android.app.Application
import android.os.CountDownTimer
import androidx.annotation.Keep
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.lunesu.pengchauferry.*
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.launch
import org.joda.time.LocalDate
import org.joda.time.LocalDateTime

class FerryViewModel(
    application: Application,
    private val ferryRepository: FerryRepository,
    private val holidayRepository: HolidayRepository
) : AndroidViewModel(application) {

    private var db: DbOpenHelper? = null

    constructor(application: Application, db: DbOpenHelper) :
        this(application,
            FerryRepository(db),
            HolidayRepository(db)
        ) {
        this.db = db
    }

    // Invoked by lazy viewModels()
    @Keep
    constructor(application: Application) : this(application, DbOpenHelper(application))

    companion object {
        fun now(): LocalDateTime =
//            if (BuildConfig.DEBUG && Utils.isEmulator)
//                LocalDateTime.now().withTime(23, 40, 0, 0)
//            else
                LocalDateTime.now()
    }

    private val countDownTimer = object : CountDownTimer(Long.MAX_VALUE, 1000) {
        override fun onTick(millisUntilFinished: Long) {
            _time.value = now()
        }

        override fun onFinish() {}
    }

    data class State(val ferries: List<Ferry>, val from: FerryPier, val day: FerryDay)

    private val _state = MutableLiveData<State>() // TODO: could be lazyMap
    private val _time = MutableLiveData(now())

    val state: LiveData<State> get() = _state
    val time: LiveData<LocalDateTime> get() = _time

    fun shouldRefresh(lastRefresh: LocalDateTime?): Boolean =
        lastRefresh == null || (today.dayOfMonth == 1 && lastRefresh.toLocalDate() != today)

    private fun shouldRefresh() : Boolean =
        holidayRepository.shouldRefresh()
            || ferryRepository.shouldRefresh()
            || shouldRefresh(Preferences(getApplication()).lastRefresh)

    private fun updateState(from: FerryPier, dow: FerryDay, autoRefresh: Boolean, filtered: Boolean) {
        var ferries = ferryRepository.getFerries(from, dow)
        if (autoRefresh && (ferries.isEmpty() || shouldRefresh())) {
            refreshAndUpdate(from)
        } else {
            if (filtered) {
                val now = _time.value?.toLocalTime()
                ferries = ferries.filter { it.time >= now }
            }
            _state.value = State(
                ferries,
                from,
                dow
            )
        }
    }

    private val today: LocalDate get() = _time.value!!.toLocalDate()

    fun getDay(date: LocalDate) =
        if (holidayRepository.getHoliday(date)) FerryDay.Holiday else FerryDay.fromDate(date)

    init {
        countDownTimer.start() // not mocked
    }

    override fun onCleared() {
        countDownTimer.cancel() // not mocked
        db?.close()
        super.onCleared()
    }

    fun toggleHoliday(): Boolean {
        val today = today
        val isHoliday = getDay(today) == FerryDay.Holiday
        holidayRepository.setHoliday(today, !isHoliday)
        _state.value?.let {
            updateState(it.from, getDay(today), true, true)
        }
        return !isHoliday
    }

    fun switchPier(pier: FerryPier) {
        updateState(pier, getDay(today), true, true)
    }

    private fun refreshAndUpdate(from: FerryPier?) = viewModelScope.launch {
        awaitAll(
            async { holidayRepository.refresh() },
            async { ferryRepository.refresh() }
        )
        Preferences(getApplication()).lastRefresh = _time.value
        from?.let {
            updateState(it, getDay(today), false, true)
        }
    }

    fun refresh() {
        refreshAndUpdate(state.value?.from)
    }

    fun fetchAll() {
        _state.value?.let {
            updateState(it.from, it.day, true, false)
        }
    }
}
